File: flutter/ios/Classes/RomanianEidSdkPlugin.swift
import Flutter
import UIKit

public class RomanianEidSdkPlugin: NSObject, FlutterPlugin {
    private var channel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?
    private var eventSink: FlutterEventSink?

    // TODO: Import RomanianEIDSDK when available
    // import RomanianEIDSDK

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = RomanianEidSdkPlugin()

        let channel = FlutterMethodChannel(name: "romanian_eid_sdk", binaryMessenger: registrar.messenger())
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)

        let eventChannel = FlutterEventChannel(name: "romanian_eid_sdk/progress", binaryMessenger: registrar.messenger())
        instance.eventChannel = eventChannel
        eventChannel.setStreamHandler(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            handleInitialize(call, result: result)
        case "isInitialized":
            handleIsInitialized(result: result)
        case "isNfcAvailable":
            handleIsNfcAvailable(result: result)
        case "readPassport":
            handleReadPassport(call, result: result)
        case "readIDCard":
            handleReadIDCard(call, result: result)
        case "scanMRZ":
            handleScanMRZ(result: result)
        case "cancelReading":
            handleCancelReading(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Method Handlers

    private func handleInitialize(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let licenseKey = args["licenseKey"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "License key required", details: nil))
            return
        }

        // TODO: Initialize with actual SDK
        // RomanianEIDSDK.shared.initialize(licenseKey: licenseKey) { success, error in
        //     if let error = error {
        //         result(FlutterError(code: "INIT_ERROR", message: error.localizedDescription, details: nil))
        //     } else {
        //         result(success)
        //     }
        // }

        // Placeholder
        result(true)
    }

    private func handleIsInitialized(result: @escaping FlutterResult) {
        // TODO: Check actual SDK initialization status
        // result(RomanianEIDSDK.shared.isInitialized)
        result(false)
    }

    private func handleIsNfcAvailable(result: @escaping FlutterResult) {
        if #available(iOS 13.0, *) {
            // TODO: Use actual SDK check
            // result(RomanianEIDSDK.shared.isNFCAvailable)
            result(true)
        } else {
            result(false)
        }
    }

    private func handleReadPassport(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let documentNumber = args["documentNumber"] as? String,
              let dateOfBirth = args["dateOfBirth"] as? String,
              let dateOfExpiry = args["dateOfExpiry"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Document number, date of birth, and date of expiry required", details: nil))
            return
        }

        // TODO: Implement actual passport reading
        // RomanianEIDSDK.shared.readPassport(
        //     documentNumber: documentNumber,
        //     dateOfBirth: dateOfBirth,
        //     dateOfExpiry: dateOfExpiry,
        //     progressHandler: { progress in
        //         self.sendProgress(progress)
        //     },
        //     completionHandler: { passportData, error in
        //         if let error = error {
        //             result(FlutterError(code: "READ_ERROR", message: error.localizedDescription, details: nil))
        //         } else if let data = passportData {
        //             result(self.passportDataToMap(data))
        //         }
        //     }
        // )

        result(FlutterError(code: "NOT_IMPLEMENTED", message: "SDK not integrated yet", details: nil))
    }

    private func handleReadIDCard(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let can = args["can"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "CAN required", details: nil))
            return
        }

        // TODO: Implement actual ID card reading
        // RomanianEIDSDK.shared.readIDCard(
        //     can: can,
        //     progressHandler: { progress in
        //         self.sendProgress(progress)
        //     },
        //     completionHandler: { idCardData, error in
        //         if let error = error {
        //             result(FlutterError(code: "READ_ERROR", message: error.localizedDescription, details: nil))
        //         } else if let data = idCardData {
        //             result(self.idCardDataToMap(data))
        //         }
        //     }
        // )

        result(FlutterError(code: "NOT_IMPLEMENTED", message: "SDK not integrated yet", details: nil))
    }

    private func handleScanMRZ(result: @escaping FlutterResult) {
        // TODO: Implement MRZ scanning
        result(FlutterError(code: "NOT_IMPLEMENTED", message: "MRZ scanning not implemented yet", details: nil))
    }

    private func handleCancelReading(result: @escaping FlutterResult) {
        // TODO: Cancel ongoing reading
        // RomanianEIDSDK.shared.cancelReading()
        result(nil)
    }

    // MARK: - Helper Methods

    private func sendProgress(_ progress: [String: Any]) {
        DispatchQueue.main.async {
            self.eventSink?(progress)
        }
    }
}

// MARK: - FlutterStreamHandler

extension RomanianEidSdkPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}

File: flutter/android/src/main/kotlin/com/eidromania/flutter/RomanianEidSdkPlugin.kt
package com.eidromania.flutter

import android.app.Activity
import android.content.Context
import android.nfc.NfcAdapter
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class RomanianEidSdkPlugin : FlutterPlugin, MethodCallHandler, ActivityAware, EventChannel.StreamHandler {
    private lateinit var channel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
    private var context: Context? = null
    private var activity: Activity? = null
    private var nfcAdapter: NfcAdapter? = null

    // TODO: Import RomanianEIDSDK when available
    // private var eidReader: EIDReader? = null

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext

        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "romanian_eid_sdk")
        channel.setMethodCallHandler(this)

        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "romanian_eid_sdk/progress")
        eventChannel.setStreamHandler(this)

        nfcAdapter = NfcAdapter.getDefaultAdapter(context)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "initialize" -> handleInitialize(call, result)
            "isInitialized" -> handleIsInitialized(result)
            "isNfcAvailable" -> handleIsNfcAvailable(result)
            "readPassport" -> handleReadPassport(call, result)
            "readIDCard" -> handleReadIDCard(call, result)
            "scanMRZ" -> handleScanMRZ(result)
            "cancelReading" -> handleCancelReading(result)
            else -> result.notImplemented()
        }
    }

    // MARK: - Method Handlers

    private fun handleInitialize(call: MethodCall, result: Result) {
        val licenseKey = call.argument<String>("licenseKey")
        if (licenseKey == null) {
            result.error("INVALID_ARGUMENTS", "License key required", null)
            return
        }

        // TODO: Initialize with actual SDK
        // eidReader = EIDReader(context!!)
        // eidReader?.initialize(licenseKey) { success, error ->
        //     if (error != null) {
        //         result.error("INIT_ERROR", error.message, null)
        //     } else {
        //         result.success(success)
        //     }
        // }

        // Placeholder
        result.success(true)
    }

    private fun handleIsInitialized(result: Result) {
        // TODO: Check actual SDK initialization status
        // result.success(eidReader?.isInitialized ?: false)
        result.success(false)
    }

    private fun handleIsNfcAvailable(result: Result) {
        val available = nfcAdapter?.isEnabled ?: false
        result.success(available)
    }

    private fun handleReadPassport(call: MethodCall, result: Result) {
        val documentNumber = call.argument<String>("documentNumber")
        val dateOfBirth = call.argument<String>("dateOfBirth")
        val dateOfExpiry = call.argument<String>("dateOfExpiry")

        if (documentNumber == null || dateOfBirth == null || dateOfExpiry == null) {
            result.error("INVALID_ARGUMENTS", "Document number, date of birth, and date of expiry required", null)
            return
        }

        // TODO: Implement actual passport reading
        // eidReader?.readPassport(
        //     documentNumber = documentNumber,
        //     dateOfBirth = dateOfBirth,
        //     dateOfExpiry = dateOfExpiry,
        //     progressHandler = { progress ->
        //         sendProgress(progress)
        //     },
        //     completionHandler = { passportData, error ->
        //         if (error != null) {
        //             result.error("READ_ERROR", error.message, null)
        //         } else {
        //             result.success(passportDataToMap(passportData))
        //         }
        //     }
        // )

        result.error("NOT_IMPLEMENTED", "SDK not integrated yet", null)
    }

    private fun handleReadIDCard(call: MethodCall, result: Result) {
        val can = call.argument<String>("can")

        if (can == null) {
            result.error("INVALID_ARGUMENTS", "CAN required", null)
            return
        }

        // TODO: Implement actual ID card reading
        // eidReader?.readIDCard(
        //     can = can,
        //     progressHandler = { progress ->
        //         sendProgress(progress)
        //     },
        //     completionHandler = { idCardData, error ->
        //         if (error != null) {
        //             result.error("READ_ERROR", error.message, null)
        //         } else {
        //             result.success(idCardDataToMap(idCardData))
        //         }
        //     }
        // )

        result.error("NOT_IMPLEMENTED", "SDK not integrated yet", null)
    }

    private fun handleScanMRZ(result: Result) {
        // TODO: Implement MRZ scanning
        result.error("NOT_IMPLEMENTED", "MRZ scanning not implemented yet", null)
    }

    private fun handleCancelReading(result: Result) {
        // TODO: Cancel ongoing reading
        // eidReader?.cancelReading()
        result.success(null)
    }

    // MARK: - Helper Methods

    private fun sendProgress(progress: Map<String, Any>) {
        activity?.runOnUiThread {
            eventSink?.success(progress)
        }
    }

    // MARK: - FlutterPlugin Lifecycle

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
    }

    // MARK: - ActivityAware

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    // MARK: - EventChannel.StreamHandler

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }
}
